### ISSUE TITLE
Losing data on _chan.exit(0) on Windows with Python >= 3.8

### ISSUE THREAD

**Issue by jerrbe** [ronf/asyncssh#586]
I am executing the [callback_math_server.py](https://github.com/ronf/asyncssh/blob/develop/examples/callback_math_server.py) example server and I am losing the "Total = x" line that is written out on line 60 of that example file when I send a break condition to end the SSH session. The session terminates before that last line of data is sent.  I was originally running with Python 3.10 on Windows 11. I also tried a Linux environment (WSL2 running Ubuntu on the same host, also with 3.10) and when executed from WSL2 the "Total = x" is correctly transmitted.  

I did have to modify the example file to remove the python3.6 requirement at the top which prompted me to try an older version of python.  Installing Python 3.6 is somewhat painful as the latest version of the cryptography module does not have pre-compiled binaries for Python 3.6 and therefore it requires the Rust compiler to install that module.  

I was able to test Python 3.7, 3.8, 3.9, 3.10, and 3.11 in a Windows 11 environment. Out of all of those 3.7 is the only one that correctly prints the Total line before the session disconnects.

I am not sure that it matters, but for all cases I was using the native ssh client in Windows 11 powershell and connecting to localhost -p 8022. I also reproduced this issue on two different Windows 11 computers.

I tried a few minor edits to the code to see if I could find a way to get the disconnect process to wait long enough to allow the data to be transmitted, however, all of my attempts failed.  I don't understand the asyncio data flow well enough to troubleshoot further.

**Comment by ronf** [comment]
I haven't really done anything with the PowerShell ssh client. In fact, I wasn't even aware that it had one... The fact that it works from other SSH clients suggests to me this might be more of an issue with the PowerShell ssh than it is with AsyncSSH.

When using the PowerShell client, how are you trying to exit out of the client? For the "Total" to be printed, you'd need to have the client send an EOF _without_ sending a CLOSE. If the client sends a CLOSE, that will close both directions and prevent the "Total" from being printed. Does PowerShell support input redirection? If so, you might try to redirect input from a file and see if the total shows up in that case, rather than trying to enter the numbers interactively.

**Comment by jerrbe** [comment]
Thank you for the feedback. 

The only way I found to actual end the session was to use "<ctrl>-c" key combination which should send a break command and the example code calls the EOF function when break is received.  I can confirm that is what happens by adding a couple of prints in the code.

I can do input redirection and that does indeed resolve the immediate problem.  For example:

```
echo 5`n4`n | ssh localhost -p 8022
Pseudo-terminal will not be allocated because stdin is not a terminal.
Enter numbers one per line, or EOF when done:
Invalid number:
Total = 9
```

The pseudo terminal line does not happen when you don't use input redirection.  

I don't know if this completely answers my question though.  I am curious how this could be a client specific problem when I can use the exact same client with a linux server running the exact same code and see totally different behavior.  And I can use the same client with an older version of python running the server and see different behavior.

I also tried one more test.  If I comment out the exit(0) line in the example code and then repeat my test using <ctrl>-c I can get the total line to print multiple times.  I believe this confirms the client is not closing the session:

```
ssh localhost -p 8022
Enter numbers one per line, or EOF when done:
5
5
6
Total = 16
6
Total = 22
Total = 22
7
Total = 29
```
where every <ctrl>-c button press results in Total being printed.

Thoughts?

**Comment by ronf** [comment]
You mentioned having to modify the callback_math_server.py example. Could you post what changes you made?

With the original version of callback_math_server.py, the "Total" is printed from within the `eof_received` callback, and that should only ever be called once. So, it doesn't match sense for that to be seen multiple times. Also, after `eof_received` is called, `data_received` should never be called again, so seeing "Total = 22" followed by it showing an input of 7 and a new Total = 29. After printing the total, it calls exit on the channel, and that should shut everything down.

I would not expect Ctrl-C to send a "break". With OpenSSH, you have to type "~" followed by "B" to send a break, where "~" is the "escape" character. Generally, Ctrl-C would trigger the client to initiate a close, which could explain it not showing the "Total" if it sent the close before the other end had a chance to process and respond to the EOF that may happen as a part of that.

Does the SSH client you are running support something like "~B"? If so, you might see if that shows "Total". It would be more likely to do so, once you really are sending a "break".

Regarding whether you have a PTY or not, that's not likely to make much of a difference other than seeing the input numbers being echoed, but there's usually an argument you can pass in to force a TTY or not regardless of stdin. In OpenSSH, adding "-tt" will do this. By getting input from a file, though, it should send EOF when it hits the end of the file, so you'd no longer have to hit Ctrl-C to get it to exit.

**Comment by jerrbe** [comment]
Yes when I run the server on linux everything works as expected. I don't have a macOS instance to test with. The only time I was observing the issue was when the server was running in Windows.

It is interesting that the "break" condition is being generated internally which is simply calling the same function that gets called with the "~B" but we get different behavior.


I did change the command that was being executed with the redirect_server example.

**Comment by ronf** [comment]
> Yes when I run the server on linux everything works as expected. I don't have a macOS instance to test with. The only time I was observing the issue was when the server was running in Windows.

Ok - I've been able to reproduce this when running the server on Windows here, even with a UNIX client. I've also been able to reproduce the difference in behavior between Ctrl-C and "\~B", even though the Ctrl-C is being handled by AsyncSSH and really should end up calling the same callback eventually. I also tried adding the `soft_eof_callback()` and it had the same behavior as Ctrl-C, suggesting that maybe there's something going on in the line editor which is somehow responsible for this.

I'll dig into this further over the weekend...

On the redirect_server example, it looks like there's a problem calling `loop.connect_write_pipe()` on the pipe gotten back from `subprocess.Popen()`:

```
Traceback (most recent call last):
  File "Y:\src\asyncssh\asyncssh\connection.py", line 1039, in _reap_task
    task.result()
  File "Y:\src\asyncssh\redirect_server.py", line 39, in handle_client
    await process.redirect(stdin=bc_proc.stdin, stdout=bc_proc.stdout,
  File "Y:\src\asyncssh\asyncssh\process.py", line 1649, in redirect
    await self._create_writer(stdin, bufsize, send_eof, recv_eof)
  File "Y:\src\asyncssh\asyncssh\process.py", line 914, in _create_writer
    await self._loop.connect_write_pipe(pipe_factory, file)
  File "C:\Program Files\Python311\Lib\asyncio\base_events.py", line 1599, in connect_write_pipe
    transport = self._make_write_pipe_transport(pipe, protocol, waiter)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\asyncio\proactor_events.py", line 681, in _make_write_pipe_transport
    return _ProactorWritePipeTransport(self,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\asyncio\proactor_events.py", line 440, in __init__
    self._read_fut = self._loop._proactor.recv(self._sock, 16)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\asyncio\windows_events.py", line 454, in recv
    self._register_with_iocp(conn)
  File "C:\Program Files\Python311\Lib\asyncio\windows_events.py", line 747, in _register_with_iocp
    _overlapped.CreateIoCompletionPort(obj.fileno(), self._iocp, 0, 0)
OSError: [WinError 6] The handle is invalid
```

I'm not sure offhand what it would take to fix this, but I'll take a look after working on the other issue. I haven't really spent any time looking into how subprocesses work on Windows.

**Comment by ronf** [comment]
Ok - I think I found the first issue. When `exit()` is called on the channel while in the middle of processing input, it later triggers an uncaught exception when it tries to write pending data in the output buffer at the bottom of `process_input()`:

```
DEBUG:asyncssh:[conn=0] Uncaught exception
Traceback (most recent call last):
  File "/Users/ronf/src/asyncssh/asyncssh/connection.py", line 1275, in data_received
    while self._inpbuf and self._recv_handler():
                           ^^^^^^^^^^^^^^^^^^^^
  File "/Users/ronf/src/asyncssh/asyncssh/connection.py", line 1522, in _recv_packet
    processed = handler.process_packet(pkttype, seq, packet)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ronf/src/asyncssh/asyncssh/packet.py", line 237, in process_packet
    self._packet_handlers[pkttype](self, pkttype, pktid, packet)
  File "/Users/ronf/src/asyncssh/asyncssh/channel.py", line 582, in _process_data
    self._accept_data(data)
  File "/Users/ronf/src/asyncssh/asyncssh/channel.py", line 413, in _accept_data
    self._deliver_data(data, datatype)
  File "/Users/ronf/src/asyncssh/asyncssh/channel.py", line 377, in _deliver_data
    self._session.data_received(decoded_data, datatype)
  File "/Users/ronf/src/asyncssh/asyncssh/editor.py", line 948, in data_received
    self._editor.process_input(data, datatype)
  File "/Users/ronf/src/asyncssh/asyncssh/editor.py", line 708, in process_input
    self._chan.write(''.join(self._outbuf))
  File "/Users/ronf/src/asyncssh/asyncssh/channel.py", line 908, in write
    raise BrokenPipeError('Channel not open for sending')
BrokenPipeError: Channel not open for sending
```

Here's a first cut at a fix:

```patch
diff --git a/asyncssh/editor.py b/asyncssh/editor.py
index d2c9ff9..174cae2 100644
--- a/asyncssh/editor.py
+++ b/asyncssh/editor.py
@@ -704,8 +704,10 @@ class SSHLineEditor:
                     self._ring_bell()

             self._bell_rung = False
-            self._chan.write(''.join(self._outbuf))
-            self._outbuf.clear()
+
+            if self._outbuf:
+                self._chan.write(''.join(self._outbuf))
+                self._outbuf.clear()
         else:
             self._session.data_received(data, datatype)
```

In cases where the break & EOF callbacks are triggered, there should be no pending output, and when there is pending output, external callbacks shouldn't be called, or if they are in the case of registering custom key handlers, it would not be expected for those to close the channel.

Interestingly, this exception was being raised on both Windows and UNIX, but for some reason it didn't prevent the "Total" line from being output in the UNIX case. I'm still not sure why that is, but this fix prevents the exception on both platforms.

**Comment by ronf** [comment]
Regarding the second issue, it looks like this should be supported when using the ProactorEventLoop, which is the default in Python 3.8 or later. See https://docs.python.org/3/library/asyncio-platforms.html#windows for more on this. However, even though the docs claim that any file-like object is supported, there are apparently issues according to https://github.com/python/cpython/issues/71019. In this case, it needs to work with pipes, but apparently it doesn't.

When I checked on the type of the pipe object, it shows up as an `<_io.BufferedWriter>` which does support `fileno()`, but `connect_write_pipe()` doesn't seem to like the handle returned from that (the value 3 in my tests).

**Comment by ronf** [comment]
The fix for the first issue is now available as commit 777d328 in the "develop" branch.

**Comment by ronf** [comment]
This fix is now available in AsyncSSH 2.14.0.
